### PR TITLE
.github: add clang-tidy workflow

### PR DESCRIPTION
**Pull request by tchaikov**
This workflow is triggered when a pull request is created targetting the "master" branch. It uses the clang-tidy tool provided by clang-tidy package to analyze the source files when building the tree.

a composite action is added so that we can add more github workflows in future. they will be able to reuse this action. a workflow using clang-include-cleaner is planed.

clang-tidy-matcher.json is added to annotate the change, so that the warnings are more visible with github webpage.

### CODE REVIEW

File: .github/workflows/clang-tidy.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+          - Debug
+          - RelWithDebInfo
+          - Dev
+          - Sanitize
+          - Coverage
+      clang_tidy_checks:
+        description: 'clang-tidy --checks options'
+        required: false
+        default: 'bugprone-use-after-move'
+        type: string
```
Comment (raphaelsc) [inline .github/workflows/clang-tidy.yaml:30]: I was running --checks=-*,bugprone-use-after-move to remove noise from default options, so we're really focusing on use-after-move only, and eventually we can incrementally enable others that we find useful.

**Comment by raphaelsc** [comment]
I have two questions:

1) will it only be a warning like codespell action?
2) will it only target files changed by PR?

**Comment by raphaelsc** [comment]
really appreciated that you worked on this, @tchaikov. clang-tidy is not perfect, but it's very good at finding use-after-move. it is certainly not perfect, as there can be false positives, but it's okay. If it's true positive, great. If it's false positive, the maintainer / PR author can mutually decide to ignore the warning.

File: .github/workflows/clang-tidy.yaml (hunk 1 above)
Comment (tchaikov) [reply .github/workflows/clang-tidy.yaml:30]: thanks. sure. added `-*`.

**Comment by tchaikov** [comment]
> I have two questions:
> 
> 1. will it only be a warning like codespell action?

we can make it so.

> 2. will it only target files changed by PR?

i think we will run it periodically. the `on.push` and `on.pull_requests` are just for testing. do you propose to run it for all PRs?

**Comment by raphaelsc** [comment]
> > I have two questions:
> > 
> > 1. will it only be a warning like codespell action?
> 
> we can make it so.
> 
> > 2. will it only target files changed by PR?
> 
> i think we will run it periodically. the `on.push` and `on.pull_requests` triggers are just for testing. do you propose to run it for all PRs?

Ideally we'd run it on every run but I am worried it can become a noise and decrease productivity. Therefore let's start with periodic runs as we discussed in the summit and then we can see how it goes. Maybe it's enough.

**Comment by denesb** [comment]
Conflicts, @tchaikov please rebase.

File: .github/actions/setup-build/action.yaml [hunk 2]
```diff
@@ -0,0 +1,110 @@
+name: setup-build-env
```
Comment (raphaelsc) [inline .github/actions/setup-build/action.yaml:1]: we need to adjust the commit message. other than that, LGTM.

Comment (tchaikov) [reply .github/actions/setup-build/action.yaml:1]: thanks for pointing this out. i updated the cover letter, but forgot to update the commit message.

**Comment by tchaikov** [comment]
v2:

- use the official PPA at https://launchpad.net/~scylladb/+archive/ubuntu/ppa 
- corrected the commit message
- use fmtlib in header only mode, so we don't link against it. jammy does not provide fmtlib 9.1 or up which is necessary for building scylla.

**Comment by tchaikov** [comment]
@scylladb/scylla-maint could you help review this change?

**Comment by tchaikov** [comment]
a successful run of the newly introduced workflow can be found at https://github.com/scylladb/scylladb/actions/runs/8918259059

**Comment by tchaikov** [comment]
the reason why it’s not triggered this time is that the .github directory is ignored on purpose.

**Review by raphaelsc — APPROVED** [review]
lgtm, thanks Kefu!

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help review this change?

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help review this change?

**Review by denesb — APPROVED** [review]

**Pull request closed without merging**